### ISSUE TITLE
Paho Java client does not support the SNI extension of TLS

### ISSUE THREAD

**Issue by miketran78727** [eclipse-paho/paho.mqtt.java#251]
Please fill out the form below before submitting, thank you!
- [x ] Bug exists Release Version 1.1.0 ( Master Branch)
- [x ] Bug exists in Snapshot Version 1.1.1-SNAPSHOT (Develop Branch)

Similar to  Python's issue https://github.com/eclipse/paho.mqtt.python/issues/11, we need to provide the support in the Java client.

**Comment by jpwsutton** [comment]
In order to implement SNI, we'll need to use Java 7 (http://docs.oracle.com/javase/7/docs/technotes/guides/security/enhancements-7.html) and above. This might be ok if we decide that the next main release is Java 7 and Upwards.

**Comment by miketran78727** [comment]
Yes, Java 7 is required.  More information how to set server name on the client side: https://www.ibm.com/support/knowledgecenter/SSYKE2_8.0.0/com.ibm.java.security.component.80.doc/security-component/jsse2Docs/sni_extension.html

**Comment by jpwsutton** [comment]
Thanks, I'll implement this into our default SSL Socket factory and add the change to our java-1.7+ branch.

For those that can't wait to the release after 1.3.0, anyone using Java 1.7 and above can implement their own SocketFactory and set it using MqttConnectOptions.setSocketFactory.
